apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/RetriableTransactionTests.cs

```diff
@@ -91,59 +92,80 @@ public async Task RetriesOnce()
+                var secondUpdateTask = connection.RunWithRetriableTransactionAsync(transaction => WorkAsync(connection, transaction, secondTaskUpdateAttempted, firstTaskUpdateAttempted, () => secondTaskCallCount++));
+
+                rowsUpdatedByTask = await Task.WhenAll(firstUpdateTask, secondUpdateTask);
             }
 
-            async Task<int> WorkAsync(SpannerConnection connection, SpannerTransaction transaction)
+            async Task<int> WorkAsync(SpannerConnection connection, SpannerTransaction transaction, ManualResetEventSlim updateEvent, ManualResetEventSlim concurrentUpdateEvent, Action callStart)
             {
-                calls++;
-
-                // Let's update some data with the retriable transaction.
-                int rows = await DmlUpdateFixtureData(key, connection, transaction);
-
-                // The first time let's update the data again using a different transaction.
-                // This will cause the retriable transaction to abort.
-                if (calls == 1)
+                // Just to count how many times this method was called.
+                // We can't have ref parameters in async methods.
+                callStart();
+                int rows;
+                try
+                {
+                    // Let's update some data with the retriable transaction.
+                    // This can fail with aborted if the other task update went in first
+                    // and this update detects the write locks already out of sync.
+                    rows = await DmlUpdateFixtureData(key, connection, transaction);
+                }
+                finally
                 {
-                    try
-                    {
-                        using (var conflictingTransaction = await connection.BeginTransactionAsync())
-                        {
-                            int conflictingRows = await DmlUpdateFixtureData(key, connection, conflictingTransaction);
-                            await conflictingTransaction.CommitAsync();
-                            Assert.Equal(2, conflictingRows);
-                        }
-                    }
-                    catch (SpannerException e)
-                    {
-                        // If there is an exception catched here we just want to propagate it,
-                        // and make the test fail. We need to wrap it so it doesn't trigger
-                        // the retry.
-                        throw new Exception("This shouldn't have happened.", e);
-                    }
+                    // We attempted an update.
+                    // Whether it succeeded or aborted, we signal the task performing the
+                    // concurrent update that we're done with our own update.
+                    updateEvent.Set();
```

Comment:
It's odd that this calls updateEvent.Set when the comment talks about signalling the *concurrent* update. Perhaps the parameters can be renamed to "thisEvent" and "otherEvent"? (Or "eventToWait" and "eventToSet"?)

Reply:
Yes, it was hard to named them, and I wasn't sure this would be clear. I'll go with thisUpdateDone and otherUpdateDone, so it will read as:
thisUpdateDone.Set() and otherUpdateDone.Wait().